### PR TITLE
Ballistics Firearm Operations Overhaul and New Improvised Firearms

### PR DESCRIPTION
**Pull request by VickiMorris**
## About The Pull Request

<details>
<summary>To-Do List for PR; Final Stages</summary>
Crossed-out checkboxes are ideas that were initially part of the PR, but have since been made defunct by changes over the month this has been written.

- [x] ~~Add .41 cal ammo types~~
- [x] Implement Pipe Rifle & Pistol
- [x] Sprites for Lever Rifle, Pipe guns, Ammo type
- [x] Craftable Med-Capacity Semi-Auto Gun / Bootleg WT-550
- [x] ~~Proper Muzzle Loading Musket / Bootleg Sniper~~
- [x] Sprites for newly mentioned firearms
- [x] **In-hand sprites** for everything
- [x] Bolt overlays for all ~~newly defined NB_BREAK and~~ PUMP action guns, to visually show state of the gun.
- [x] Crafting recipes
- [x] Refactor rifle loading code
- [x] ~~Modify NO_BOLT (break action guns) code to function similar to Mosins for loading/unloading~~
- [x] Modify Pump Action guns to function similar to the Mosin for cycling cartridges
- [x] Clean up and streamline code.
- [x] Balance balance balance ammo and firearms (Might take a bit)

</details>

The overall goal of this PR is to make the aquisition of firearms and ammunition by antagonists (and by extension, crew and security in dire circumstances) easier, by providing more (lower quality) firearms and ammunition to be crafted. Paired with this is a series of tweaks to existing firearms to change, what I feel, are oversights or just poor design choices. This is the third re-write of this PR about section, so I will keep it to a bulleted list and not turn it into a 'used-car salesman' pitch.

---
- Two new firearms, a single shot break-action rifle chambered in 7.62 (same ammo type as the Mosin), and an open-bolt semi-auto 'carbine' chambered in 9mm (used by the Stechkin APS, Saber SMG, and Uzi).
- A lever action rifle; 8+1 'pump'-action gun chambered in .38, currently unobtainable and a byproduct of previous testing.
- Makeshift ammo types for 7.62, 9mm, 10mm, .38, .357, and a new type of shotgun ammo.
- Improvised shotgun ammo now has better range; each pellet has a random range of '3 to 8' instead of '1 to 8' 
- ~~The improvised shotgun can no longer chamber 'high-power' rounds; this includes buckshot, lethal slugs, and all tech shells asides from dragon's breath. All other ammo still function normally.~~
- The improvised shotgun has a 10% chance to explode when firing 'high-power' rounds (Buckshot, Slugs, and all Tech Shells asides from dragons breath), increasing by 5% each time a high-power round is fired from it. After suffering enough 'barrel stress' (30 stress / 6 shots) from firing high-power rounds, there is a 5% chance that low-power rounds will also cause a backfire. An additional variable has been included to allow admins and mappers to create improvised shotguns that ignore this mechanic.
- Slightly more complex crafting recipes for all improvised guns, requiring a couple extra components and a welder to construct.
- Improvised weapons are no longer loaded when crafted, requiring ammo to be made separately. They no longer require or spawn with a removable firing pin either.
- Refactoring of the Mosin rifle code, allowing it to be applied to other guns. The Sniper now uses this bolt type and fires marginally slower as a result.
- The ability to half-cock 'PUMP' bolt weapons (most shotguns). The ability to directly chamber rounds in PUMP and TWO_STEP bolt weapons while the bolt is locked open. This applies to most shotguns, the Mosin, and the Sniper Rifle.
- Hacksaws, obtainable from hacked autolathes; it can be used as a slower working and lower damage substitute to medical circular saws. This includes surgery, sawing weapons, and butchering.
- Sawing weapons actually increases the recoil of guns by 1 now, instead of setting the recoil to 1. Almost all sawn-off guns have higher recoil as a result.
- Replacing spent ammunition in a magazine has been entirely gutted. Magazines loaded with spent ammo must be emptied before they can be reloaded; this mainly effects NO_BOLT type guns (Break-Action Shotguns and Revolvers)
- Ammo boxes and removable magazines are all larger, now size category SMALL instead of TINY. This is to cut down on the *absurd* amount of ammo that can be carried for magazine fed guns.
- Removable magazines are no-longer filled instantly from ammo boxes, loaded at a rate of 1 bullet per 4 deciseconds. This can be done while moving, if reloading magazines on the go is needed. They can no longer pick up bullets from the floor like regular ammo boxes can either.
- Guns no longer chamber a round during their initialization, and must be manually cycled before firing for the first time. They still spawn with the same amount of ammo. An exception is made for NO BOLT guns, as there is no way to do this without giving them less ammo.
- The admin-only trait "NICE_SHOT" has been expanded upon. In addition to granting a wider ricochet and auto-aim degree, it allows the user to rack PUMP and TWO_STEP firearms with a single hand. This will likely never see use, but allows the creation of a supernaturally strong gunslinger through admin shenanigans.

<details>
<summary>For the nitty-gritty statistics, please click here.</summary>

- Pipe Carbine
Bulky-Sized firearm, can be stored on Belt slot only. Has a fire rate of 2 bullets per second, and an innate spread of 10 degrees. Chambered in 9mm (20 base damage).
The craftable magazines can hold 9 bullets, and can be upgraded with matter bins; each tier lets it hold 3 additional bullets, upto a max of 18 bullets with a bluespace matter bin.
- Pipe Rifle
Bulky-Sized firearm, no innate item slot. Can have a sling made from 10 cable coil to allow it to be carried on your Back and can be sawn-off, similar to the improvised shotgun. Chambered in 7.62 (40 base damage).

New Ammo Types
---
```
- 9mm (Improv); 20 damage, 10 innate spread
- 10mm (improv); 27 damage, 10 innate spread
- .38 (Improv); 25 damage, 1 ricochet (no auto-aim), 10 innate spread
- .357 (Improv); 50 damage, 15 innate spread, -10 AP
- 7.62 (Improv); 27-30 damage (randomized), 0 AP
- 7.62 (Hotload); 32-35 damage, 15 AP, '0.7 speed' (decisecond per tile, base is 0.8)
- Shotgun (Glasspack); 5 pellets, 15 degree cone. 6 damage per pellet (30 point-blank)
Max range of 8 tiles, 0.5 damage drop-off per pellet per tile.
80% chance for each pellet to 'shrapnel' and embed (60% embed chance).
```

</details>

---
Sprites for the lever action rifle were derived from the original lever action shotgun sprite, as made by OceanFish1 for #3153
Sawing noise as used for the hacksaw is derived from https://freesound.org/people/larryblag/sounds/333288/ by larryblag, originally released under Creative Commons 0.

## Why It's Good For The Game
The second half of this PR is easy enough to justify, the options for obtaining firearms as an antag at the moment are to order it from cargo (essentially steal it from security), actually steal it from security, purchase it from an uplink (BBs, Heretics, Changelings, Cult, etc stay losing), get Gamer Loot as a shaft miner or explorer, or craft the single improvised gun that's available. Having more options for antags to work their way towards better gear is good, especially when ranged options for most are pretty weak.

This doesn't directly change how Standard, Locking, and Open bolt guns function, but indirectly tweaks them with some changes to how magazines function; a bulk of this overhaul looks at changing the Pump bolt type to allow half-cocking and directly chambering round (functioning as a half-automatic version of the Mosin's bolt type), and defining a new bolt type (and reclassifying a couple of firearms in the process).

This is to make certain logical choices more favorable; having a gun with an larger internal magazine is better because I can fire more before needing to reload (Improv Shotgun is weaker than Bartender's Shotgun is weaker than Combat Shotgun) and reloading takes time. Having multiple magazines for my gun means I have more ready-to-use firepower in exchange for actual ammo capacity (Carrying multiple loaded magazines vs. one magazine and multiple boxes of ammo).

The slight logic changes to TWO_STEP and PUMP firearms allows more tactical usage of them; being able to chamber rounds directly is a buff to shotguns and the sniper, as you can now choose your next shot without needing to mess around with loading ammo into the magazine and then racking to charge the gun. While this isn't as much of an issue with shotguns, I found it miserable to actually load any of the specialty ammo into the sniper rifle without having a spare mag for the two bullets you get in each specialty ammo box.

## Testing Photographs and Procedure

<details>
<summary>Screenshots&Videos</summary>
The lever action rifle and pipe pistol are both derivative sprites from the lever-action shotgun and improvised shotgun, respectively. The ammo box fills out the boxes of ammo with the 4th color (others being red, blue, and green), the actual .41 caliber cartridges are based off of the Mosin / 7.62mm cartridge, and the improvised ones looking like a slimmed down improvised shotgun shell. Since removing the .41 caliber cartridge, the lever rifle is now chambered in .38 (same as the Det's revolver) and the new ammo box and proper metal cartridge sprites have been removed.

The original lever-action shotgun and improvised shotgun are off to the right (on the white tiles), and all the new sprites are on a blue background for better contrast.
image

The pipe SMG / pipe repeater now exists! A slightly more expensive to make full auto (but slow firing) open bolt mag fed firearm (magazine not included in crafting). It uses the *very powerful* 9mm cartridge (at 20 damage per shot), has mild spread, and an upgradable magazine (initially stores 9 rounds, can be upgraded to 18 with a T4 matter bin)
image

General screenshot showing the remaining new items, and some of the new bolt overlays used for shotguns to represent being locked open (to make it easier to recognize at a glance)
image

New inhand sprites, used for the improvised shotgun and pipe rifle (and their respective sawn-off variants )
custominhands

Recipes as of time of marking this as ready for review
gun recepies
ammo recepies

Videos below document some of the latest changes, and guns in their current state in the PR.

https://github.com/BeeStation/BeeStation-Hornet/assets/97719613/49b92c3a-6a6f-4e70-8d7b-d0fbc593212a

https://github.com/BeeStation/BeeStation-Hornet/assets/97719613/8a1b08fe-9358-484d-9818-473c31a20cc7

This video documents the changes to the improvised shotgun, and detonation/barrel stress mechanic while firing high-power rounds.

https://github.com/BeeStation/BeeStation-Hornet/assets/97719613/4bf38da9-b2b7-4d08-b9e8-def64dac9623

</details>

## Changelog
:cl: Impish_Delights, OceanFish1
add: New craftable a7.62, shotgun, 9mm, 10mm, .38, and .357 ammo types. Lever action rifle, craftable pipe rifle and pipe pistol, craftable pipe SMG and magazines.
add: Two-step bolt guns (Sniper & Mosin) can have ammo directly chambered, allowing loading a single shot without a magazine.
add: Pump-action guns can be half-cocked, locking them open to allow directly chambering rounds as well.
add: Hacksaws are now available from hacked autolathes, very slow manual saw for surgery and cutting down guns. The Russian Smuggler hunter shuttle spawns with one as well.
soundadd: Sawing noise for use with new hacksaw item. Original sound by larryblag on freesound.org
imageadd: Sprites for new ammo types, guns and accessories. Custom in-hand sprites for improv shotgun and pipe rifle. Original lever action shotgun sprite made by OceanFish1.
balance: Improvised shotguns have an incremental chance to explode when firing high-power ammo (buckshot, slugs, most tech shells). No danger with improv ammo and all other ammo types (incen, dart, rubbershot, beanbag, improv, breacher, dragonsbreath).
balance: The Sniper Rifle now uses the same bolt type as the Mosin Nagant. Fire rate is marginally slower as a result, and requires two interactions to cycle a new cartridge.
balance: Sawing adds 1 recoil instead of setting recoil to 1. Sawn-off shotguns have higher recoil across the board as a result. Slightly lowers Improv shotgun's base recoil as it would be absurd otherwise.
balance: The derringer has weapon weight 0.5 instead of 1, and no longer has hard-defined unwielded spread to bring it in line with the changes from high speed gun accuracy
balance: Improvised guns no longer require firing pins, and longer give free ammo when crafted.
balance: Improvised shotgun shells have better minimum range.
balance: All ammo boxes and external magazines are size SMALL instead of size TINY now. Magazines now load 1 bullet at a time, and no longer vacuum up bullets.
tweak: Due to rising safety concerns, guns are no longer shipped with a chambered round, and must be manually cycled before use.
tweak: Improvised weapons require an igniter, and a welder to construct now.
tweak: The admin-only trait "NICE_SHOT" now allows the user to cycle PUMP and TWO_STEP bolt firearms with a single hand, instead of requiring both.
refactor: Changes mosin/boltaction rifle code to be a new bolt type instead of bespoke code for the rifle parent type.
/:cl:

### CODE REVIEW

**Comment by PestoVerde322** [comment]
Don't do those changelogs as much memeable it is, also strongly suggest to make a custom icon for this new weapon instead of recycling the nagant icon.

Also add an image to show what they are!

**Comment by VickiMorris** [comment]
This is very much a WIP draft I slapped together in the 2 minutes I had before work, I'll keep it in mind for the future though. I plan in adding new inhand sprites and other stuff, but it's just a very rough draft as I hone in everything.

**Comment by PowerfulBacon** [comment]
I dislike the idea of having to pump the shotgun twice ngl, I would honestly just prefer if it flicked an animation state and had a cooldown

**Comment by PowerfulBacon** [comment]
In general the changes make sense but I feel like they are going to make guns way more clunky and annoying to use, in the end its going to be more punishing to new players trying to get the gun into a state where it isn't rejecting your ammo while it won't slow experienced players down that much. If balancing is the intention with the ammo insertion changes then its a balance change which significantly affects newer players over experienced ones.

**Comment by VickiMorris** [comment]
This is very much a work in progress on the asset side of things (asides from making assets for completely new items) and the plan is to have item states / overlays that represent when the gun is is in loading mode vs firing mode. The changes to pump actions aren't hard to revert, and I could just keep the required manual ejection of casings on NO_BOLT guns so you're still required to dump before reloading (instead of mashing a speedloader against a fully loaded cylinder and magically teleporting out just the spent casings). I still think the changes to TWO_STEP / BOLT_ACTION (Mosin, now Sniper rifle) are a net positive though.

**Comment by VickiMorris** [comment]
PUMP bolt guns and NO_BOLT guns are back to how they used to function. As a temporary hacky way for just testing with, PUMP invokes a callback timer that drops the bolt after `rack_delay` if the gun is still wielded. I'll change it back to just, full cycling the chamber later, and keep loading directly into the chamber as an auxiliary feature.

**Comment by DarnTheMarn** [comment]
> in the end its going to be more punishing to new players trying to get the gun into a state where it isn't rejecting your ammo

I don't know what you mean. I don't see a situation where the gun would be rejecting your ammo for no reason? If it's not open, no ammo goes in aside from tube fed shotguns. Mosins already work this way, you can't slam ammo into it when the bolt is closed. It would also have a cool effect of being a small nerf to the .357 people complain about.

>I dislike the idea of having to pump the shotgun twice ngl, I would honestly just prefer if it flicked an animation state and had a cooldown

Its more consistent with other guns like the mosin where you have to activate it twice before it shoots. IMO one should be changed to two actions, or the other should be changed to one action as well. Cycling a pump can be done in one fluid motion, so can operating a bolt on a bolt action.

**Comment by VickiMorris** [comment]
Upsides and downsides of each bolt type, as of current commit
---
Put as a spoiler because there's a lot of text down below, we've got seven different bolt types baybeeeee
<details>

Standard & Locking:
```
+(Typically) Removable magazine
+Quick cycling, one step
+Auto-ejects spent casings while cycling
+Retains chambered round when magazine is removed
-Requires a magazine to chamber rounds

~Only difference between standard and locking is that locking 'locks' open after firing its last shot
~Untested, but you should be able to chamber directly into a locking firearm's
  chamber while it is locked open (most automatic pistols)
```

Open:
```
+Can be cocked with or without a magazine
+Removable magazine
+Auto-ejects spent casings while cycling
+Quick reload speed (remove mag, insert new mag, can be pre-cocked)
-Cannot retain ammo in the chamber, must have a magazine to function
```

Pump:
```
+Can load directly into battery while open
+Can load magazine at any time
+Auto-ejects spent casings while cycling
~Two-step cycling, but second-step is automatic
~Can half-cycle to lock open and load into chamber
-Slow reload, single bullet at a time only
```

Two Step:
```
+Can load directly into battery while open
+Can multiload magazine
+Auto-ejects spent casings while cycling
-Slow cycling, two steps
-Can only load magazine while open/unable to fire
```

No Bolt:
```
+Can fire entire 'magazine' before reloading
+Can load and unload entire 'magazine' at once (Dumps all ammo, speed loaders)
~Moderately quick reload, efficiency scales with capacity (Dump, shove bullets in, can immediately fire)
-Must manually eject spent cartridges
-Can only eject entire 'magazine' at once, not just spent casings
```

No Bolt Break:
```
+Can fire entire 'magazine' before reloading
+Can load and unload entire 'magazine' at once (Dumps all ammo, speed loaders)
-Slow reload, three steps (Open, Dump/Load, Close)
-Can only load 'magazine' while open
-Must manually eject spent cartridges
-Can only eject entire 'magazine' at once, not just spent casings
```
</details>

**Comment by VickiMorris** [comment]
Removing the complex NO BOLT design, it's too much of a headache and I'm having second thoughts on how it would mesh with BeeStation as a whole.

**Comment by VickiMorris** [comment]
Will return to this to fix the conflict this week, work has been a drag.

**Comment by VickiMorris** [comment]
I'm just not sure where I'd slip the lever action in for obtainability, as a 7+1 .38 special 'pump-action' rifle isn't super strong, but there's nothing it really fits with flavor-wise in terms of jobs. If anything, I'd probably look at a "wild west outlaw" syndie bundle, or *maybe* a ruins.

I did toy around with the idea of a Wild West Sheriff hunter ship for the fugatives game mode, but I do not want to feature-creep this PR even more.

File: code/modules/research/designs/autolathe_designs.dm
Comment (Rukofamicom) [inline code/modules/research/designs/autolathe_designs.dm]: Why does the hacksaw take glass to produce?

File: code/modules/projectiles/ammunition/ballistic/shotgun.dm
Comment (Rukofamicom) [inline code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: This is purely a suggestion and not a forced requirement for merging:

Instead of makeshift/shitty shotguns not being allowed to use high_power rounds, would you consider making them have a chance to explode and harm the user if they try to use high powered rounds?

**Review by Rukofamicom — COMMENTED** [review]
Just a couple of questions before approval

File: code/modules/research/designs/autolathe_designs.dm
Comment (VickiMorris) [reply code/modules/research/designs/autolathe_designs.dm]: IIRC I based the cost of the hacksaw off of the circular saw, tbh it does make more sense for it to be just pure iron instead of iron and glass. I'll tweak that to be 12000 iron instead of 9000 iron and 3000 glass when I get a chance.

File: code/modules/projectiles/ammunition/ballistic/shotgun.dm
Comment (VickiMorris) [reply code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: That'd probably be the better solution instead of a hard-ban on high-power shells. I'm not sure if I'd prefer it to be a flat percent chance (probably around 15%) each shot, or if high-power ammo 'damages' the barrel and builds up a malfunction chance (with it guaranteed to explode after 5 or so shots).

File: code/modules/research/designs/autolathe_designs.dm
Comment (VickiMorris) [reply code/modules/research/designs/autolathe_designs.dm]: Updated to be 12000 iron instead.

File: code/modules/projectiles/ammunition/ballistic/shotgun.dm
Comment (DarnTheMarn) [reply code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: Apologies but I have to do the cliché and state my unsought opinion. I hate RNG based things like that, I myself would prefer if there was a set number of shots before it blew up. Its more fair that way.

Comment (Rukofamicom) [reply code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: Strongly disagree - RNG absolutely deserves a place in combat within an RP oriented setting, especially as a limiter against some of the more powerful tools available. It would not deserve a place if this were a heavily pvp oriented server but this isn't the case.

Somewhere between 10% and 25% is a good rate - it's low enough that they can be used at least somewhat reliably, but high enough that malfunctions are a threat and prevent them from being a totally dominating factor.

File: code/modules/research/designs/autolathe_designs.dm
Comment (Rukofamicom) [reply code/modules/research/designs/autolathe_designs.dm]: 👍

File: code/modules/projectiles/ammunition/ballistic/shotgun.dm
Comment (DarnTheMarn) [reply code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: I do think RNG in something like this where it only effects a person who gets greedy is far less egregious than say the tonfa baton PR where I said the same thing as that had the chance to just instantly sleep someone you used it on.

Comment (VickiMorris) [reply code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: Just added a ~~bullshit RNG~~ stress mechanic to improvised shotguns. 10% base chance to explode (similar to det's revolver) that slowly goes up as high-power ammo is fired through it (+5% per shot). After enough strain is put on the gun (+30% / 6 shots), there's a slim chance (5%) for low-power ammo to cause a detonation as well. The reinforced variable is used to ignore this mechanic now.

Video is under testing evidence. Feels pretty alright in the little bit of testing I did with it tbh. There are pop-ups at certain stress thresholds, and it gives a vague idea of the current stress level in the examine prompt.

Comment (Rukofamicom) [reply code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: It's almost eerie how that matches the compromise I was going to suggest number for number but got interrupted before I could finish typing it on lunch

**Comment by Rukofamicom** [comment]
Need to review most recent code changes properly at PC but this is as good as approved if I don't see issues with those

File: code/modules/projectiles/ammunition/ballistic/shotgun.dm
Comment (DarnTheMarn) [reply code/modules/projectiles/ammunition/ballistic/shotgun.dm:8]: Is this the extremely rare compromise that makes everyone equally happy instead of equally unhappy? I think it might be.

**Comment by VickiMorris** [comment]
Glasspack rounds against an unarmored target with a full-auto shotgun are comical. I will be doing a quick change to address the stupid amount of debris/trash this makes, as currently *every single pellet that has shrapnel will leave behind a piece of shrapnel as an object.*

image

File: code/modules/projectiles/guns/ballistic/shotgun.dm
Comment (Rukofamicom) [inline code/modules/projectiles/guns/ballistic/shotgun.dm]: `The barrel is warped and cracked` would look better for the final message imo. The current wording assumes they have been examining it as it worsens, which will not always be the case. 

Additionally these thresholds don't match up with the warning messages very well right now: 
Thresholds should be `<10` and `<20` rather than `== 0` and `<= 25`.

**Review by Rukofamicom — CHANGES_REQUESTED** [review]

File: code/modules/projectiles/guns/ballistic/shotgun.dm
Comment (VickiMorris) [reply code/modules/projectiles/guns/ballistic/shotgun.dm]: Updated to have stuff line up. Barrel becomes warped at 10 stress (when message occurs), and cracked at 25 stress now.

After 2 shots, it's warped (20% misfire). After 5 shots, it's cracked (35% misfire).

**Review by Rukofamicom — APPROVED** [review]

**Comment by Rukofamicom** [comment]
Secondary approval by bacon given via Discord

**Pull request merged**
